Extended const eval engine to const-evaluate a chain of tuple_element_addr insts

mhong opened the pull request:
One use case is to deabstract a const tuple used in the `kernelSize` and `strides` attrs of the TF `maxPoolV2` op.

One example is given by the SIL snippet below. There we set
`elementsAtInit[index]` to %178 with `index` being 3. To const-evaluate the last
copy_addr below, we first const-evaluate %191, which const-evaluates %179, which
involves const-evaluating %183, where the writer to that tuple elt address is
"copy_addr %114". Eventually it const-evaluates %101 to fill in the vlaue for
%183. The continues until the writer insts of all the other tuple elements of
%179 are const-evaluated, yielding a const tuple for %179. This finally gives us
the const value at address %178.

```
%101 = tuple_extract %96 : $(Int32, Int32, Int32, Int32), 3
%108 = alloc_stack $Int32
store %101 to %108 : $*Int32
%114 = tuple_element_addr %110 : $*(Int32, Int32, Int32, Int32), 3
%121 = tuple_element_addr %110 : $*(Int32, Int32, Int32, Int32), 3
copy_addr [take] %108 to [initialization] %121 : $*Int32
copy_addr %114 to [initialization] %183 : $*Int32
%177 = integer_literal $Builtin.Word, 3
%178 = index_addr %130 : $*Int32, %177
%179 = alloc_stack $(Int32, Int32, Int32, Int32)
%183 = tuple_element_addr %179 : $*(Int32, Int32, Int32, Int32), 3
%191 = tuple_element_addr %179 : $*(Int32, Int32, Int32, Int32), 3
copy_addr [take] %191 to [initialization] %178 : $*Int32
```

Also refactored the code as follows:
- Renamed computeXXX() methods to getXXX(), when such methods use the
  cache (`calculatedValues`).

- Wrote comments to contrast `getSingleWriterAddressValue()` with
  `getConstAddrAndLoadResult()`, and corrected `decodeAllocUninitializedArray()`
  logic to call `getSingleWriterAddressValue()` instead of
  `getConstAddrAndLoadResult()`.

- Moved more code from `getConstantValue()` to `computeConstantValue()`, so that the
  former becomes a thin wrapper for the latter, and the wrapper only handles
  caching.


<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-8413](https://bugs.swift.org/browse/SR-8413).

Commit:
Extendec const eval engine to const-evaluate a chain of tuple_element_addr insts.

One example is given by the SIL snippet below. There we set
`elementsAtInit[index]` to %178 where index being 3. To const-evaluate the last
copy_addr below, we first const-evaluate %191, which const-evaluates %179, which
involves const-evaluating %183, where the writer to that tuple elt address is
"copy_addr %114". Eventually it const-evaluates %101 to fill in the vlaue for
%183. The continues until the writer insts of all the other tuple elements of
%179 are const-evaluated, yielding a const tuple for %179. This finally gives us
the const value at address %178.

```
%101 = tuple_extract %96 : $(Int32, Int32, Int32, Int32), 3
%108 = alloc_stack $Int32
store %101 to %108 : $*Int32
%114 = tuple_element_addr %110 : $*(Int32, Int32, Int32, Int32), 3
%121 = tuple_element_addr %110 : $*(Int32, Int32, Int32, Int32), 3
copy_addr [take] %108 to [initialization] %121 : $*Int32
copy_addr %114 to [initialization] %183 : $*Int32
%177 = integer_literal $Builtin.Word, 3
%178 = index_addr %130 : $*Int32, %177
%179 = alloc_stack $(Int32, Int32, Int32, Int32)
%183 = tuple_element_addr %179 : $*(Int32, Int32, Int32, Int32), 3
%191 = tuple_element_addr %179 : $*(Int32, Int32, Int32, Int32), 3
copy_addr [take] %191 to [initialization] %178 : $*Int32
```

mhong commented:
@swift-ci please test tensorflow

mhong commented:
I'm not very happy with the structure of this patch, and would love to receive some feedback.

mhong commented:
@dingobye as FYI, who also worked on const eval.

Commit:
Polished the code a bit.

Commit:
Addressed feedback from @dingobye at https://github.com/apple/swift/pull/18563

bgogul commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -1078,20 +1080,44 @@ ConstExprFunctionState::computeSingleStoreAddressValue(SILValue addr) {
+/// This does the heavy-lifting of computeSingleStoreAddressValue() above, and
+/// may recursively call itself when const-evaluating TupleElementAddrInst.
+SymbolicValue
+ConstExprFunctionState::computeSingleStoreAddressValueHelper(SILValue addr) {
```
May not pertinent to this CL, but would this be better implemented as a SILValue/SILInstructionVisitor?

mhong replied:
Yes this also came up in some past PRs from @lattner on const eval.

lattner replied:
This function seems highly redundant with the existing getConstant() support for addresses.  Can we reuse it?

bgogul commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -1183,6 +1209,59 @@ ConstExprFunctionState::computeSingleStoreAddressValue(SILValue addr) {
+                         [](SymbolicValue v) { return v.isConstant(); }))
+          break;
+        // Otherwise we keep going, hoping to find more users of the tuple inst
+        // that write into other tuple elements.
```
Is this the right thing to do? Can't we have a use of the tuple before all values are initialized?

mhong replied:
Good point on dealing with multiple writers. I added a `writeOnlyOnce` param, to make sure we won't have multiple writers to the same agg element.

Not sure what'd be a swift test case that can trigger "multiple writers". From our discussion, please feel free to follow up on this. :)

Regarding "use of a tuple before all values are initialized", I believe SILGen cannot generate such code.

bgogul replied:
Thanks for the clarifications. 

Actually, the existing implementation of  computeSingleStoreAddressValue returns unknown in such cases. See the case for StoreInst:
`    if (auto *si = dyn_cast<StoreInst>(user)) {
       if (use->getOperandNumber() == 1) {

         // If we have already found a value for this stack slot then we're done:                                                                                                                                                                                               
         // we don't support multiple assignment.                                                                                                                                                                                                                               
         if (memoryObject->getValue().getKind() != SymbolicValue::UninitMemory)
           return evaluator.getUnknown(addr, UnknownReason::Default);
        ...
`
Should we have the same behavior here?

mhong replied:
Yes. the code below should do it
```c++
if (failed)
    return evaluator.getUnknown(addr, UnknownReason::Default);
```

Commit:
Addressed more feedback from @dingobye.

Commit:
Added a `writeOnlyOnce` param to `updateIndexedElement()` based on input from @bgogul.

lattner commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -1078,20 +1089,44 @@ ConstExprFunctionState::computeSingleStoreAddressValue(SILValue addr) {
+      return tupleEltAddr;
+    }
+  }
   if (!addrInst)
```
This approach predated the current approach to handling addresses in the constexpr stuff.  Instead of pattern matching the SIL instructions directly, how about just calling:

auto addrCst = getConstantValue(addr);

and then handling the various forms of address constants?  That will allow handling recursive forms of indexing.

One other random point: generally you shouldn't call computeConstantValue directly, because this skips caching.  Is there a reason you're not calling getConstantValue?

mhong replied:
I studied the design and code base more, and came to the belief that we'll need two different code paths for handling the (fairly small) set of inst types that return address values, such as `pointer_to_address`, `index_addr` and `tuple_element_addr`. 

e.g. given an inst like
```
%132 = tuple_element_addr %131 : $*(Int32, Int32, Int32, Int32), 0
```

When we are reading from `%132`, we should const-evaluating %131 into a tuple const first, and use that result to const-eval `%132`. But if we are writing into `%132` (as handled by the code here), we will be const-evaluating a writer of `%132`, as a way to const-evaluate `%131`. The flow of information differs between these two cases. Fortunately, the chunk of code in `getConstValue()` that handles these addr-typed insts is fairly small, so there is not much added complexity.

For more context, PTAL the comment block above `getSingleWriterAddressValue()`, and how it contrasts with `getConstAddrAndLoadResult()` (I renamed `computeLoadResult()` to the latter, which is based on `getConstantValue()`.

lattner reviewed (commented):
Thanks for tackling this, but it would be great to see if you could reuse the existing address analysis logic.  Thanks!

Commit:
Merge branch 'tensorflow' of github.com:apple/swift into const_eval_tuple_elts

Commit:
- Adressed feedback by @latter and refactored the code:

- Eliminated getSingleWriterAddressValueHelper()

- Renamed computeXXX() methods to getXXX(), when they use when they use the
  cache (`calculatedValues`).

- Wrote comments to contrast getSingleWriterAddressValue() with
  getConstAddrAndLoadResult(), and corrected decodeAllocUninitializedArray()
  logic to call getSingleWriterAddressValue() instead of
  getConstAddrAndLoadResult().

- Moved more code from getConstantValue() to computeConstantValue(), so that the
  former becomes a thin wrapper for the latter, and the wrapper only handles
  caching.

mhong commented:
@swift-ci please test tensorflow

mhong commented:
Thanks for the feedback. I refactored the code and am much happier with the new structure. PTAL.

Commit:
Polished code and comments.

mhong commented:
@swift-ci please test tensorflow

lattner commented on lib/SILOptimizer/Mandatory/TFConstExpr.cpp:
```diff
@@ -187,7 +188,19 @@ static void lookupOrLinkWitnessTable(ProtocolConformanceRef confRef,
+  // single store value.  Since this is a very different computation, split it
+  // out to its own path.
+  if (!fn && value->getType().isAddress() && isa<AllocStackInst>(value)) {
+    SymbolicValue result = getSingleWriterAddressValue(value);
```
LG, you can simplify the code to just "return getSingleWriter..." of course.

lattner reviewed (approved):
very nice, thank you for adjusting this!

Commit:
Addressed feedback.

mhong commented:
Thanks for the review and feedback!
